cmd/promoted-image-governor/main.go

```diff
@@ -0,0 +1,314 @@
+	if f := fs.Lookup("kubeconfig"); f != nil {
+		f.Usage = kubeconfigFlagDescription
+		// https://i.kym-cdn.com/entries/icons/original/000/018/012/this_is_fine.jpeg
+		defer func() { opts.kubeconfig = f.Value.String() }()
+	} else {
+		fs.StringVar(&opts.kubeconfig, "kubeconfig", "", kubeconfigFlagDescription)
+	}
+	fs.StringVar(&opts.ciOperatorconfigPath, "ci-operator-config-path", "", "Path to the ci operator config")
+	fs.BoolVar(&opts.dryRun, "dry-run", true, "Whether to run the controller-manager with dry-run")
+	fs.Var(&opts.ignoredImageStreamTagsRaw, "ignored-image-stream-tags", "A regex to match tag in the form of namespace/name:tag format. Can be passed multiple times.")
+	fs.StringVar(&opts.releaseControllerMirrorConfigDir, "release-controller-mirror-config-dir", "", "Path to the release controller mirror config directory")
+	if err := fs.Parse(os.Args[1:]); err != nil {
+		logrus.WithError(err).Fatal("could not parse input")
+	}
+	return opts
+}
+
+func (o *options) validate() error {
+	if o.ciOperatorconfigPath == "" {
+		return fmt.Errorf("--ci-operator-config-path must be set")
+	}
+	if o.releaseControllerMirrorConfigDir == "" {
+		return fmt.Errorf("--release-controller-mirror-config-dir must be set")
+	}
+	for _, s := range o.ignoredImageStreamTagsRaw.Strings() {
+		re, err := regexp.Compile(s)
+		if err != nil {
+			return fmt.Errorf("failed to compile regex from %q: %w", s, err)
+		}
+		o.ignoredImageStreamTags = append(o.ignoredImageStreamTags, re)
+	}
+	return nil
+}
+
+func addSchemes() error {
+	if err := imagev1.AddToScheme(scheme.Scheme); err != nil {
+		return fmt.Errorf("failed to add imagev1 to scheme: %w", err)
+	}
+	return nil
+}
+
+func validatePromotedTags(tags []api.ImageStreamTagReference) error {
+	tagCount := map[api.ImageStreamTagReference]int{}
+	for _, tag := range tags {
+		if _, ok := tagCount[tag]; ok {
+			tagCount[tag] = tagCount[tag] + 1
+		} else {
+			tagCount[tag] = 1
+		}
+	}
+	var dups []string
+	for tag, c := range tagCount {
+		if c > 1 {
+			dups = append(dups, tag.ISTagName())
+		}
+	}
+	if len(dups) > 0 {
+		return fmt.Errorf("found tags promoted more than once: %s", dups)
+	}
+	return nil
+}
+
+func tagsToDelete(ctx context.Context, client ctrlruntimeclient.Client, promotedTags []api.ImageStreamTagReference, toIgnore []*regexp.Regexp, imageStreamRefs []ImageStreamRef) ([]api.ImageStreamTagReference, error) {
+	imageStreamsWithPromotedTags := map[ctrlruntimeclient.ObjectKey]interface{}{}
+	for _, promotedTag := range promotedTags {
+		imageStreamsWithPromotedTags[ctrlruntimeclient.ObjectKey{Namespace: promotedTag.Namespace, Name: promotedTag.Name}] = nil
+	}
+
+	tagsToCheck := map[api.ImageStreamTagReference]interface{}{}
+	var errs []error
+	for objectKey := range imageStreamsWithPromotedTags {
+		imageStream := &imagev1.ImageStream{}
+		if err := client.Get(ctx, objectKey, imageStream); err != nil {
+			if !kerrors.IsNotFound(err) {
```

Comment:
Why do we want to special-case this? Streams that get promoted to should exist

Reply:
We do have those cases when I run the tool locally.
Added logs for the case.